### ISSUE TITLE
AdaptiveTrackSelection stays on the highest variant

### ISSUE THREAD

**Issue by ChernyshovYuriy** [google/ExoPlayer#2660]
Dear colleagues,

Please correct my approach if I provide it in a wrong way.
The task is to do dynamic selection within provided video variants on the throttled network.
I use DASH clear content with many variants available:
```
http://bitmovin-a.akamaihd.net/content/sintel/sintel.mpd
```
I slightly modify ExoPlayer Sample app.
1. TrackSelectionHelper.java
    Make modifier **public**
```
public void setOverride(int group, int[] tracks, boolean enableRandomAdaptation)
```
2. PlayerActivity.java
```
@Override
  public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
    if (playbackState == ExoPlayer.STATE_ENDED) {
      showControls();
    }
    // Hardcode these lines only for provided content in order to test functionality
    if (playbackState == ExoPlayer.STATE_READY) {
      trackSelectionHelper.setOverride(0, new int[]{0,1,2,3,4,5,6,7,8}, false);
    }
    updateButtonVisibilities();
  }
```
Then, I use Android Emulator, because it allows to provide different types / conditions of the Cellular network:
`Nexus 5 API 25 Android 7.1.1`

Test case:
- Use emulator's "Extended Controls" view to manage Cellular network ("..." icon right bottom)
- Network Type **Full**, Signal Strength **Good**
- Once video started it switched to highest variant and play
- Change network to Type **GPRS**, Signal Strength **Moderate**
- Observe video stoped on the highest variant and State is Buffering for a long time, until buffer is full, then it play buffered data and behave the same way again

**Note** that if next line excluded (back to default code):
```
trackSelectionHelper.setOverride(0, new int[]{0,1,2,3,4,5,6,7,8}, false);
```
and with the provided test case steps, video plays well selecting dynamically appropriate variant.

To sum up, whether my approach of dynamic selection is wrong or is there a bug in ExoPlayer?
My thoughts was that I am able to provide a range of variants to be dynamic within, for instance:
```
new int[]{0,1,2,3,4,5,6,7,8}
```
is a full list of variants.
So if, I provide:
```
new int[]{0,1,2,3,4}
```
I expected that ExoPlayer must be dynamic within those 4 variants.
But my test case shows different behaviour.

**Comment by WeiChungChang** [comment]
Hi @ChernyshovYuriy
 

> Use emulator's "Extended Controls" view to manage Cellular network ("..." icon right bottom)
Network Type Full, Signal Strength Good
Once video started it switched to highest variant and play
Change network to Type GPRS, Signal Strength Moderate
Observe video stoped on the highest variant and State is Buffering for a long time, until buffer is full, then it play buffered data and behave the same way again

Observe video stopped on the highest variant and State is Buffering **_for a long time_**, until buffer is full, then it play buffered data and behave the same way again

I guess it is because selectTracks will clean up all buffered data.

Since its default value is "Default" but now you set it item-by-item to all; 
**_It clean up the buffered data and re-fetch again._**
So you will experience a long buffering pending.

Please trace the flow by

1. selectTracks() of DashMediaPeriod.java
2. release() of ChunkSampleStream.java
3. disable() of DefaultTrackOutput.java
4. finally, clearSampleData() will clean up all buffered data. 

**_The only thing I wonder is where you trigger the override to take effect?_**

It is, the override of TrackSelectionHelper takes effect only by 
**_public void onClick(DialogInterface dialog, int which)_** 
So I guess you must either trigger it manually or press any button meanwhile(?).

Thanks

**Comment by ojw28** [comment]
I think your approach is wrong. Why are you setting the override every time the player reports that it's ready, as opposed to just setting it once?

The ExoPlayer demo app does quite a good job of demonstrating the ability to select a subset of tracks for adaptation. You can select the subset you want under the "Video" button in the demo. I'd suggest you take a look at the calls being made there.

**Comment by ChernyshovYuriy** [comment]
@ojw28 , I use this simple approach jut to show you that using **public API** I am not able to get dynamic variants selection. While ExoPlayer does an excellent dynamic selection using approach introduced in Sample App, usage of this particular public API (set a range of variants instead of only one, like it was prior to version 2.X) does not provide expected behavior.
Meanwhile, suppose I need to limit user to use selected (selected by the client application) range of variants, how can I do it then? This question cover a functionality of providing a strict limit of variant to the client application, for instance something like set "min" and "max" bandwidth at runtime.

**Comment by ojw28** [comment]
I'm quite confused about what you're trying to do. What is the difference between what you're trying to do and what the demo app successfully does? Or are you saying that the feature is broken even in the demo app?

**Comment by ChernyshovYuriy** [comment]
Ok, let's consider simple use case.
There are 8 [0, ..., 7] video variant available in provided example.
Suppose I need to limit player to use only middle range, let's say [2,3,4,5]
Or, i need to limit player to play only lowest ones [0,1,2].
Previously, in version 1.5.x, I was able to set a concrete variant from the list.
Since 2.x, public API provide a way to specify a set of variants rather then only one.
I was expected that player are dynamic within provided range.
Unless I am fundamental wrong with my approach of limiting a player to stay on the selected range.

**Comment by ChernyshovYuriy** [comment]
@ojw28 , I am seeing ExoPlayer example does it in a different way.
Let me try to apply logic that your example does. Thanks for bringing my attention to sample code once again.
I will provide feedback as soon as possible.

**Comment by ChernyshovYuriy** [comment]
Ok, here is update.
Without any code changes, and given test scenario provided originally, when select all variants like shown in the image attached, once variant reached its highest value (6M) and network started throttled, the variant stays on the 6M for a long time with a state "buffering".
The question is - do I misunderstand a concept of dynamic variants selection within a range or is there a bug in ExoPlayer. From the perspective of the ExoPlayer API user, it seems like unexpected behaviour.

![device-2017-04-10-093107](https://cloud.githubusercontent.com/assets/931955/24864349/ee5667e4-1dd1-11e7-9498-43020e52487b.png)

**Comment by ojw28** [comment]
I think the problem is that the "GPRS moderate" profile appears to be way too bad for pretty much anything. With that profile enabled, it took >4 minutes to load a single fairly simply web page in Chrome.

For ExoPlayer, it appears that the requests time out due to the poor quality of the connection. Under these conditions our bandwidth estimate doesn't get updated because we're not successfully loading any data from which an estimate can be derived. I don't think it's really important what happens, however. Playback of any variant is going to fail under conditions this poor.

**Comment by ChernyshovYuriy** [comment]
Going to close this issue because track selector works as expected.
Thank you guys for all input provided!

P.S.
Whoever is wondering why this ticket was open - I tested dynamic selection on LecoTV, device that requires high bandwidth to play HD and UHD content.
